Plutarch/Verify.hs

```diff
@@ -0,0 +1,277 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-unticked-promoted-constructors #-}
+{-# OPTIONS_GHC -Wredundant-constraints #-}
+
+module Plutarch.Verify (
+  PTryFrom (PExcess, ptryFrom),
+  pcheckType,
+  pcheckByteStr,
+  pcheckInt,
+  pcheckList,
+  pcheckMap,
+) where
+
+import Plutarch.Builtin (
+  PAsData,
+  PBuiltinList,
+  PBuiltinMap,
+  PBuiltinPair,
+  PData,
+  PIsData (pfromData),
+  pasByteStr,
+  pasInt,
+  pasList,
+  pasMap,
+  pdata,
+  pforgetData,
+  pfstBuiltin,
+  ppairDataBuiltin,
+  psndBuiltin,
+ )
+
+import Plutarch.ByteString (PByteString)
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal.Other (
+  POpaque,
+  PType,
+  Term,
+  pcon,
+  perror,
+  pforce,
+  phoistAcyclic,
+  plam,
+  plet,
+  (#),
+  type (:-->),
+ )
+import Plutarch.Unit (PUnit (PUnit))
+
+import Plutarch.DataRepr.Internal (
+  PDataRecord,
+  PLabeledType ((:=)),
+  pdcons,
+  pdnil,
+ )
+
+import Plutarch.Bool (pif, (#==))
+
+import Plutarch.List (
+  phead,
+  pmap,
+  ptail,
+ )
+
+import Plutarch.Unsafe (punsafeBuiltin, punsafeCoerce)
+import qualified PlutusCore as PLC
+
+import Plutarch.TermCont (TermCont, tcont, unTermCont)
+
+{- |
+    Note: PAsData POpaque ~ PData
+-}
+
+----------------------- The class PTryFrom ----------------------------------------------
+
+{- |
+    This checks the datastructure for validity.
+    If you don't care about parts of the datastructure
+    don't verify those parts, just let it return a PData
+    instead
+    Be aware this might get really expensive, so only
+    use it if you cannot establish trust otherwise
+    (e.g. via only checking a part of your Data with
+    PTryFrom)
+-}
+class PTryFrom (a :: PType) (b :: PType) where
+  --     "safest" type --^            ^-- target type
+  type PExcess a b :: PType
+  ptryFrom :: Term s a -> TermCont s (Term s b, Term s (PExcess a b))
+
+----------------------- PData instances -------------------------------------------------
+
+instance PTryFrom PData (PAsData PInteger) where
+  type PExcess PData (PAsData PInteger) = PInteger
+  ptryFrom opq = do
+    ver <- tcont $ plet (pasInt # opq)
+    pure $ (punsafeCoerce opq, ver)
+
+instance PTryFrom PData (PAsData PByteString) where
+  type PExcess PData (PAsData PByteString) = PByteString
+  ptryFrom opq = do
+    ver <- tcont $ plet (pasByteStr # opq)
+    pure $ (punsafeCoerce opq, ver)
+
+instance PTryFrom PData PData where
+  type PExcess PData PData = PUnit
+  ptryFrom opq = pure $ (opq, pcon PUnit)
+
+instance {-# OVERLAPPING #-} PTryFrom PData (PAsData (PBuiltinList PData)) where
+  type PExcess PData (PAsData (PBuiltinList PData)) = PBuiltinList PData
+  ptryFrom opq = do
+    ver <- tcont $ plet (pasList # opq)
+    pure $ (punsafeCoerce opq, ver)
+
+instance
+  {-# OVERLAPS #-}
+  ( PTryFrom PData (PAsData a)
+  , PTryFrom PData (PAsData b)
+  ) =>
+  PTryFrom PData (PAsData (PBuiltinMap a b))
+  where
```

Comment:
This instance is something to be careful of IMO - in many cases the efficient approach would be instead of verifying the whole structure 'outside in' by checking the whole structure, only to use parts of it, checking 'inside out' by first breaking down the structure then checking the relevant parts.

Reply:
i.e. fusing the verification and processing I suppose. We could reintroduce the whole deep-shallow schism. Not sure it's worth it.

Reply:
Yeah, exactly.

Reply:
I don't know whether the shallow thing would be feasible in this case tho, because after verifying that the outer type is a `PBuiltinList`, you end up with a list of `PBuiltinPair` (*not* `PAsData BuiltinPair`) which you can't verify standalone (at least I don't know of a way to do it).